### PR TITLE
[SPARK-9672][MESOS] Don’t include SPARK_ENV_LOADED when passing env vars

### PR DESCRIPTION
**Pull request by pashields**
This contribution is my original work and I license the work to the project under the project's open source license.

### CODE REVIEW

**Comment by tnachen** [comment]
Can we not filter anything MESOS_ as well? Users set the MESOS_NATIVE_JAVA_LIBRARY in spark_env.sh and this could filter this out

**Comment by pashields** [comment]
Thanks for the review @tnachen 

Anything that is in spark-env.sh shouldn't be filtered now that SPARK_ENV_LOADED isn't being sent (since it will be re-sourced on the driver side). I added some code to make sure MESOS_* vars are passed through anyways so that they could be set elsewhere.

**Comment by andrewor14** [comment]
@tnachen any follow ups on this?

**Comment by tnachen** [comment]
This LGTM

**Comment by andrewor14** [comment]
ok to test

File: core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala [hunk 1]
```diff
@@ -413,6 +412,15 @@ private[spark] object RestSubmissionClient {
...
     val conf = new SparkConf
-    run(appResource, mainClass, appArgs, conf)
+    val env = filterSystemEnvironment(sys.env)
+    run(appResource, mainClass, appArgs, conf, env)
+  }
+
+  /**
+   * Filter non-spark environment variables from any environment.
+   */
+  def filterSystemEnvironment(env: Map[String, String]): Map[String, String] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: please make this private

File: core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala [hunk 2]
```diff
@@ -413,6 +412,15 @@ private[spark] object RestSubmissionClient {
...
+    val sparkVars = env.filter { case (k, _) => k.startsWith("SPARK_") || k.startsWith("MESOS_") }
+    sparkVars - "SPARK_ENV_LOADED"
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: this makes a copy of the map, can you just push this into the filter?
```
env.filter { case (k, _) =>
  (k.startsWith("SPARK_") && k != "SPARK_ENV_LOADED) || k.startsWith("MESOS_")
}
```

Comment (pashields) [inline core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: Done

File: core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala (hunk 1 above)
Comment (pashields) [inline core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: There is a test for it, so I'd need to make it visible to the test somehow or remove the test.

I've changed it to protected[rest]. LMK if you have an alternate solution or would like me to make it private and remove the test.

**Comment by pashields** [comment]
Thanks for the review @andrewor14.

File: core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala [hunk 3]
```diff
@@ -413,6 +412,16 @@ private[spark] object RestSubmissionClient {
...
     val conf = new SparkConf
-    run(appResource, mainClass, appArgs, conf)
+    val env = filterSystemEnvironment(sys.env)
+    run(appResource, mainClass, appArgs, conf, env)
+  }
+
+  /**
+   * Filter non-spark environment variables from any environment.
+   */
+  protected[rest] def filterSystemEnvironment(env: Map[String, String]): Map[String, String] = {
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: this can be strictly private, I've fixed that on merge.

Comment (pashields) [inline core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: Thanks for helping to guide this through. Apologies for lack of Scala knowledge.

Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/deploy/rest/RestSubmissionClient.scala]: no worries, hope to see more patches from you!

**Pull request closed without merging**